### PR TITLE
TLS part 1: Enable TLS if cert is available

### PR DESCRIPTION
**Pull request by sed-i**
Tracking issue: canonical/cos-lite-bundle#75 

This PR enables TLS for alertmanager when a certificate is available.

This involves:
- adding `tls-certificates` to metadata
- writing private key and certificate to disk
- writing a web config file
- updating the pebble command with the web config file

This PR goes in tandem with https://github.com/canonical/observability-libs/pull/49.

TODO:
- [x] Scale to two units and make sure both units' TLS endpoints are reachable
- [x] Remove relation and:
  - [x] make sure TLS endpoint is NOT reachable
  - [x] re-relate, and make sure TLS endpoint _is_ reachable
- [x] Confirm TLS endpoint still reachable after an upgrade sequence
- [x] Think what to do with `insecure_skip_verify` in this line in alertmanager charm code: `"global": {"http_config": {"tls_config": {"insecure_skip_verify": True}}},`

### CODE REVIEW

**Comment by lucabello** [comment]
I've addressed some of the missing pieces:
* about the certificate files deletion, I've added it conditionally since `container.push()` overwrites the file anyway; this way we don't re-create certificates every time the configuration is updated
* the **reload** logic has been moved to a separate method so it can be called in the order you mentioned was correct

File: lib/charms/observability_libs/v0/cert_manager.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+        # We're in "relation-joined", so the relation should be there
+        peer_relation = self._peer_relation
+
+        # Just in case we already have a private key, do not overwrite it.
+        # Not sure how this could happen.
+        # TODO figure out how to go about key rotation.
+        if not self._private_key:
+            private_key = generate_private_key()
+            self._private_key = private_key.decode()
```
Comment (lucabello) [inline lib/charms/observability_libs/v0/cert_manager.py:142]: `generate_private_key()` should be called with the `private_key_password`!

File: lib/charms/observability_libs/v0/cert_manager.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        """Generate the private key and store it in a peer relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        # We're in "relation-joined", so the relation should be there
+        peer_relation = self._peer_relation
+
+        # Just in case we already have a private key, do not overwrite it.
+        # Not sure how this could happen.
+        # TODO figure out how to go about key rotation.
```
Comment (lucabello) [inline lib/charms/observability_libs/v0/cert_manager.py:139]: I would maybe add some utility function that does this; wherever the input for key rotation comes from (a charm action?):

```python
def rotate_keys(self, _) -> None:
    """Generate a new private key and certificate for key rotation, revoking the old one."""
    if not self.charm.unit.is_leader() or not self._peer_relation:
        return
        
    private_key = generate_private_key(self.private_key_password)
    self._private_key = private_key.decode()
    
    self._certificate_revoked()
    self._generate_csr()
```

Generate a new key, revoke the current certificate and ask for a new one seems reasonable

**Review by lucabello — COMMENTED** [review]
`CertManager` has been changed to provide certificate-per-app (i.e., only the leader can generate a CSR); however, the previous scope was certificate-per-unit.

Are we saying that all units should share the same certificate?

**Comment by sed-i** [comment]
Tested successfully with:
```yaml
# # Deploy bundle
# juju deploy --trust ./bundle.yaml
#
# # Obtain alertmanager IP address
# IPADDR=$(juju status --format json am | jq -r '.applications.am.units."am/0".address')
#
# # Make sure alertmanager web-external-hostname is locally routable
# echo "$IPADDR alertmanager.local" | sudo tee -a /etc/hosts
#
# # Make sure charm code created web-config, cert and key files
# juju ssh --container alertmanager am/0 ls /etc/alertmanager/
#
# # Inspect server cert and confirm `X509v3 Subject Alternative Name` field is as expected
# echo | openssl s_client -showcerts -servername $IPADDR:9093 -connect $IPADDR:9093 2>/dev/null | openssl x509 -inform pem -noout -text
#
# # Save CA cert locally
# juju show-unit am/0 --format yaml | yq '.am/0."relation-info"[0]."local-unit".data.ca' > /tmp/cacert.pem
#
# # Confirm alertmanager TLS endpoint reachable
# curl --fail-with-body --capath /tmp --cacert /tmp/cacert.pem https://alertmanager.local:9093/-/ready

---
bundle: kubernetes
applications:
  am:
    charm: ./alertmanager-k8s_ubuntu-20.04-amd64.charm
    series: focal
    scale: 1
    trust: true
    resources:
      alertmanager-image: ubuntu/prometheus-alertmanager:0.23-22.04_beta
    options:
      web_external_url: https://alertmanager.local
  ca:
    charm: self-signed-certificates
    channel: edge
    scale: 1
relations:
- [am:certificates, ca:certificates]
```

**Comment by sed-i** [comment]
Now testing with this bundle (with traefik):
```yaml
# # Deploy bundle
# juju deploy --trust ./bundle.yaml
#
# # Obtain IP addresses
# IPADDR=$(juju status --format json trfk | jq -r '.applications.trfk.address')
# IPADDR0=$(juju status --format json am | jq -r '.applications.am.units."am/0".address')
# IPADDR1=$(juju status --format json am | jq -r '.applications.am.units."am/1".address')
# IPADDR2=$(juju status --format json am | jq -r '.applications.am.units."am/2".address')
#
# # Make sure traefik external-hostname is locally routable
# echo "$IPADDR cluster.local" | sudo tee -a /etc/hosts
#
# # Make sure charm code created web-config, cert and key files
# juju ssh --container alertmanager am/0 ls /etc/alertmanager/
#
# # Inspect server cert and confirm `X509v3 Subject Alternative Name` field is as expected
# echo | openssl s_client -showcerts -servername cluster.local -connect cluster.local 2>/dev/null | openssl x509 -inform pem -noout -text
#
# # Save CA cert locally
# juju show-unit am/0 --format yaml | yq '.am/0."relation-info"[0]."local-unit".data.ca' > /tmp/cacert.pem
#
# # Confirm traefik ingress has `https` for alertmanager's server url
# juju ssh --container traefik trfk/0 cat /opt/traefik/juju/juju_ingress_ingress_6_am.yaml
#
# # Confirm alertmanager TLS endpoint reachable
# curl --fail-with-body --capath /tmp --cacert /tmp/cacert.pem https://cluster.local/tlstest-am-0/-/ready

---
bundle: kubernetes
applications:
  am:
    charm: ./alertmanager-k8s_ubuntu-20.04-amd64.charm
    series: focal
    scale: 3
    trust: true
    resources:
      alertmanager-image: ubuntu/prometheus-alertmanager:0.23-22.04_beta
    options:
      web_external_url: https://alertmanager.local
  ca:
    charm: self-signed-certificates
    channel: edge
    scale: 1
  trfk:
    charm: traefik-k8s
    channel: edge
    scale: 1
    options:
      external_hostname: cluster.local
relations:
- [am:certificates, ca:certificates]
- [trfk:certificates, ca:certificates]
- [am:ingress, trfk]
```

File: lib/charms/observability_libs/v0/cert_manager.py [hunk 3]
```diff
@@ -345,4 +345,8 @@ def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEven
     def _on_certificates_relation_broken(self, event: RelationBrokenEvent) -> None:
         """Clear the certificates data when removing the relation."""
         if self._peer_relation:
-            pass # TODO clear/delete the peer relation
+            if private_key := self._private_key:  # preserve the private key
+                self._peer_relation.data[self.charm.unit].clear()
```
Comment (sed-i) [inline lib/charms/observability_libs/v0/cert_manager.py:349]: @lucabello shouldn't we clear `event.unit`  instead of `self.charm.unit`?

Comment (PietroPasotti) [reply lib/charms/observability_libs/v0/cert_manager.py:349]: you can't clear the databag of another unit, only your own

Comment (sed-i) [reply lib/charms/observability_libs/v0/cert_manager.py:349]: Ok so not `clear()`. How about `del` or `pop`?
We don't need to clear the remote data bag in the juju sense, just in the ops data structure, for this particular hook.
The following line,
```
self._peer_relation.data[self.charm.unit].update({"private_key": private_key})
```
is quite brittle: we'd rather not manually recreate reldata to schema.

File: src/charm.py [hunk 4]
```diff
@@ -92,6 +92,20 @@ def __init__(self, *args):
         super().__init__(*args)
         self._stored.set_default(config_hash=None, launched_with_peers=False)
 
+        if url := self.model.config.get("web_external_url"):
+            extra_sans_dns = [cast(str, urlparse(url).netloc)]
+        else:
+            extra_sans_dns = None
```
Comment (Abuelodelanada) [inline src/charm.py:98]: Just a matter of taste ;-) 

```suggestion
        extra_sans_dns = None
        if url := self.model.config.get("web_external_url"):
            extra_sans_dns = [cast(str, urlparse(url).netloc)]
```

File: src/charm.py [hunk 5]
```diff
@@ -206,12 +221,14 @@ def __init__(self, *args):
...
-            }
-        ]
+        job = {
+            "metrics_path": f"{external_url.path}/metrics",
+            "static_configs": [{"targets": [f"{external_url.hostname}:{external_url.port}"]}],
+        }
+        if self.server_cert.cert:
+            job["scheme"] = "https"
+
+        return [job]
```
Comment (Abuelodelanada) [inline src/charm.py:231]: `external_url.hostname` is provided by `Traefik`, for instance:

imagen


On the other hand `self.server_cert.cert` is provided by `CertManager`.

If `Traefik` doesn't support `HTTPS` in the `URLs` it provides, we can end up generating an invalid URL

File: lib/charms/observability_libs/v0/cert_manager.py (hunk 3 above)
Comment (sed-i) [reply lib/charms/observability_libs/v0/cert_manager.py:344]: Hmm:
```
TypeError: 'RelationData' object does not support item deletion
```
```
AttributeError: 'RelationData' object has no attribute 'pop'
```

File: src/charm.py [hunk 6]
```diff
@@ -583,18 +639,39 @@ def _common_exit_hook(self) -> None:
...
+
+    def _push_tls_files(self):
+        if key := self.server_cert.key:
+            self.container.push(self._key_path, key, make_dirs=True)
+        else:
+            self.container.remove_path(self._key_path, recursive=True)
+        if cert := self.server_cert.cert:
+            self.container.push(self._server_cert_path, cert, make_dirs=True)
+        else:
+            self.container.remove_path(self._server_cert_path, recursive=True)
```
Comment (Abuelodelanada) [inline src/charm.py:673]: Does this method name represents well what this method actually does?
What about `_update_tls_files` instead?

File: src/charm.py [hunk 7]
```diff
@@ -649,13 +726,29 @@ def _on_update_status(self, _):
...
+            #  written, is this really necessary?
+            return yaml.safe_dump(yaml.safe_load(raw_yaml_str))
+
+        config_state = ConfigFileSystemState()
+        # add all configuration paths to the state
+        for path in self._config_paths:
+            content = _normalize(self.container.pull(self._config_path))
+            config_state.add_file(path, content)
+
+        return hash(config_state)
```
Comment (Abuelodelanada) [inline src/charm.py:745]: We decided to avoid sub-functions a while ago...

File: src/config_utils.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+import hashlib
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Union
+
+from ops.model import Container
+
+
+def _sha256(hashable) -> int:
```
Comment (Abuelodelanada) [inline src/config_utils.py:15]: As far as I understand `hashable` is either `str` or `bytes`.

Shouldn't we have this signature?

```suggestion
def _sha256(hashable: Union[str, bytes]) -> int:
```

File: src/charm.py (hunk 5 above)
Comment (sed-i) [reply src/charm.py:231]: Yep. And I think we said self-scrape is going to be FQDN?
Anyway, let's leave this for #165? I added your comment [there](https://github.com/canonical/alertmanager-k8s-operator/issues/165#issuecomment-1603307584).

File: src/charm.py (hunk 6 above)
Comment (sed-i) [reply src/charm.py:673]: Fixed in next commit.

File: src/charm.py (hunk 7 above)
Comment (sed-i) [reply src/charm.py:745]: Meh :D

File: src/config_utils.py (hunk 8 above)
Comment (sed-i) [reply src/config_utils.py:15]: Fixed in next commit.

File: src/charm.py [hunk 9]
```diff
@@ -77,11 +76,8 @@ class AlertmanagerCharm(CharmBase):
     _server_cert_path = "/etc/alertmanager/alertmanager.cert.pem"
     _key_path = "/etc/alertmanager/alertmanager.key.pem"
 
-    _stored = StoredState()
```
Comment (PietroPasotti) [inline src/charm.py:89]: :100:

**Review by PietroPasotti — COMMENTED** [review]
Not enthusiastic about the name "WorkloadManager". It really is about syncing multiple config files, I wish it was clear from the name. 
I'll play with merging it with the config builder somehow

**Review by PietroPasotti — APPROVED** [review]
It's too large to really pass with a fine comb, but what I see is good and ITWT (in testing we trust)

File: src/workload_manager.py
[the remark this answers is not in the document]
Comment (sed-i) [reply src/workload_manager.py:1]: Also: investigate how we could generalize the workload manager to have it reusable elsewhere. @lucabello

File: src/alertmanager.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    def check_config(self) -> Tuple[str, str]:
+        """Check config with amtool.
+
+        Returns stdout, stderr.
+        """
+        if not self.is_ready:
+            raise ContainerNotReady(
+                "cannot check config: alertmanager workload container not ready"
+            )
+        proc = self._container.exec(["/usr/bin/amtool", "check-config", self._config_path])
```
Comment (Abuelodelanada) [inline src/alertmanager.py:152]: I would extract `/usr/bin/amtool` to a constant, so it is easy to find in the future.

File: src/config_builder.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
+# Copyright 2021 Canonical Ltd.
```
Comment (Abuelodelanada) [inline src/config_builder.py:1]:
```suggestion
# Copyright 2023 Canonical Ltd.
```

**Review by Abuelodelanada — APPROVED** [review]

File: src/alertmanager.py (hunk 10 above)
Comment (sed-i) [reply src/alertmanager.py:152]: Fixed in next commit.

File: src/config_builder.py (hunk 11 above)
Comment (sed-i) [reply src/config_builder.py:1]: Fixed in next commit.

**Pull request merged**